org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java

```diff
@@ -503,9 +502,30 @@ public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.I
 				.forEach(inMemoryCu -> {
 					pathToUnit.put(new String(inMemoryCu.getFileName()), inMemoryCu);
 				});
-
-		// note that this intentionally overwrites an existing working copy entry for the same file
-		pathToUnit.put(new String(sourceUnit.getFileName()), sourceUnit);
+		
+		if (!pathToUnit.keySet().isEmpty() && sourceUnit instanceof BasicCompilationUnit) {
+			// BasicCompilationUnit sometimes just has the last segment of the path,
```

Comment:
Should we also check that in the condition, eg add some `Path.of(fileName),getNameCount() == 1`? Could such a check replace the `instanceof BasicCompilationUnit` and thus make the code more self-explanatory?